synapse/storage/schema/main/delta/75/01_add_profiles_full_user_id_column.sql.postgres

```diff
@@ -0,0 +1,21 @@
+/* Copyright 2023 The Matrix.org Foundation C.I.C
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ALTER TABLE profiles ADD COLUMN full_user_id TEXT;
+
+-- Add a new constraint on the new column, mirroring the `profiles_user_id_key`
+-- constraint.
+ALTER TABLE ONLY profiles
+    ADD CONSTRAINT profiles_full_user_id_key UNIQUE (full_user_id);
```

Comment:
I think if you do it this way, you don't have to do things differently for both Postgres and SQLite:

```sql
CREATE UNIQUE INDEX profiles_full_user_id_key ON profiles (full_user_id);
```

(I actually didn't know you *could* do it your way!)

(Though I believe for your version you could also `ALTER TABLE profiles ADD COLUMN full_user_id TEXT UNIQUE;`, SQLite isn't happy with that though. Presumably doing it at creation time means it doesn't have to do a full table scan to check for uniqueness.)


However `CREATE UNIQUE INDEX` involves a full table walk, so I think we'd want to do this in the background....

Even creating a unique column in Postgres directly takes a lot of time :/. I don't think we can avoid doing the background index update, sorry...
```
synapse=> \timing
Timing is on.

synapse=> begin;
synapse=*> alter table events add column wombats TEXT UNIQUE;
ALTER TABLE
Time: 6789.267 ms (00:06.789)
synapse=*> abort;


synapse=> begin;
synapse=*> alter table events add column wombats TEXT;
ALTER TABLE
Time: 0.911 ms
synapse=*> abort;
```

Reply:
So I think?? I have done what you asked, and chose to build both indexes in the background for consistency.